### PR TITLE
make self referential AST fast too

### PR DESCRIPTION
**Pull request by DKormann**
## problem
building of self referential computation graph leads to exponential runtime even without being realized:

```
from tinygrad.tensor import Tensor
from tinygrad.ops import Device,BinaryOps

Device.DEFAULT = "METAL"

N=30
p = LazyBuffer.fromCPU(np.array([1]))
for _ in range(N): p = p.e(BinaryOps.ADD, p)

```

## source:
when building the Ast a list of exponentially many buffers is created in order to set .child property on all buffers.

## solution:

create a function to get unique buffers for LazyOp. just calling set() on the normal buffers doesn't help since we still would visit the .src exponentially often.

### CODE REVIEW

**Comment by geohot** [comment]
Why is buffers even needed? Is there a way to do this refactor that loses lines instead of gains them?

**Comment by DKormann** [comment]
it seems connected to all kinds of slow downs

**Comment by DKormann** [comment]
wow your right actually buffers was basically not needed.

**Comment by geohot** [comment]
Now minimize the diff. Do you have to change the name?

**Comment by DKormann** [comment]
could make the buffers into property like before but its 2 + lines or so I think not worth it?

File: tinygrad/lazy.py [hunk 1]
```diff
@@ -143,8 +143,7 @@ def key(self):
 
   def _device_extra_args(self) -> Dict[str, str]: return {"device": self.device.split(":", 1)[1]} if ":" in self.device else {}
 
-  @property
-  def buffers(self) -> Tuple[LazyBuffer, ...]: return (self,)
+  def buffers(self,_=None) -> Tuple[LazyBuffer, ...]: return (self,)
```
Comment (geohot) [inline tinygrad/lazy.py:146]: Why can't this stay a property?

Comment (DKormann) [reply tinygrad/lazy.py:146]: I can even put it as a regular property in the __init__ since it always just returns itseld

**Comment by DKormann** [comment]
the whole hash thing also doesn't seem relevant when caching buffers.
super sorry this took so long

File: tinygrad/ops.py [hunk 2]
```diff
@@ -1,8 +1,10 @@
 from __future__ import annotations
 import importlib, inspect, functools, pathlib, re
 from enum import Enum, auto
+
```
Comment (geohot) [inline tinygrad/ops.py:4]: Why change this whitespace?

File: tinygrad/ops.py [hunk 3]
```diff
@@ -52,13 +54,11 @@ class LazyOp:
...
   def __repr__(self): return f"LazyOp(op={self.op}, src={self.src}, arg={self.arg})"
-  @property
-  def buffers(self):
-    buffers: Tuple[Union[LazyOp, LazyBuffer], ...] = ()
-    try:  # NOTE: the linearizer's key function maps the buffers to ints, and LOCAL_BUFFER is used. we don't care about buffers in these cases
-      for x in self.src: buffers += x.buffers
-    except AttributeError: buffers = ()
-    return buffers
+  @functools.cached_property
+  def buffers(self)-> Tuple[LazyBuffer,...]: return tuple(dedup(sum([x.buffers for x in self.src],())))
```
Comment (geohot) [inline tinygrad/ops.py:58]: `) -> `

File: tinygrad/ops.py [hunk 4]
```diff
@@ -52,13 +54,11 @@ class LazyOp:
...
+  @functools.cached_property
+  def hash(self): return hash((self.op,self.src, self.arg))
+  def __hash__(self):return self.hash
```
Comment (geohot) [inline tinygrad/ops.py:61]: `: return`

**Comment by geohot** [comment]
Looks good, just formatting issues now. Ask, is this beautiful code?

**Pull request merged**